pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +182,28 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, file_obj):
+        def do_next_upload(checksum, size):
+            data = file_obj.read(self._CHUNK_SIZE)
+            if data:
+                checksum.update(data)
+                return f_flat_map(
+                    self._do_upload(data, upload_id, size),
+                    lambda _: do_next_upload(checksum, size + len(data)),
+                )
+            else:
+                # nothing more to upload, return checksum and size
+                if is_string:
+                    file_obj.close()
```

Comment:
I think the file closing should be more robust, seems there are some error cases where file_obj.close may never be called.

I recommend trying a plain old future callback which will be invoked whether the future succeeds or fails.

So, the last few lines of _do_upload_file may be:

```
out = f_flat_map(f_return(), lambda _: do_next_upload(hashlib.sha256(), 0))
if is_string:
  # When upload completes, close the file we opened (whether or not it succeeded)
  out.add_done_callback(lambda _: file_obj.close())
return out
```

Reply:
Yeah, makes sense. I didn't think about unexpected failures... guess that's one reason why it's called 'unexpected'.